api.go

```diff
@@ -72,6 +76,46 @@ type FileMatch struct {
 	Version string
 }
 
+// ChunkMatch is a set of non-overlapping matches within a contiguous range of
+// lines in the file.
+type ChunkMatch struct {
+	// Content is a contiguous range of complete lines that fully contains Ranges.
+	Content []byte
+	// ContentStart is the location (inclusive) of the beginning of content
+	// relative to the beginning of the file. It will always be at the
+	// beginning of a line (Column will always be 1).
+	ContentStart Location
+
+	// Ranges is a set of matching ranges within this chunk. Each range is relative
+	// to the beginning of the file (not the beginning of Content).
+	Ranges []Range
+
+	// FileName indicates whether this match is a match on the file name, in
+	// which case Content will contain the file name.
+	FileName bool
+
+	Score      float64
+	DebugScore string
+}
+
+type Range struct {
+	// The inclusive beginning of the range.
+	Start Location
+	// The exclusive end of the range.
+	End Location
+
+	SymbolInfo *Symbol
+}
+
+type Location struct {
+	// 0-based byte offset from the beginning of the file
+	ByteOffset int
+	// 1-based line number from the beginning of the file
+	LineNumber int
+	// 1-based column number (in runes) from the beginning of line
+	Column int
```

Comment:
A lot of this information seems like it can be hydrated in since you have `ChunkMatch.Content`. What is the benefit of making it part of the API here? Do you plan on making Content optional? Or is this for just ergonomics.

Just being overly sensitive to size of things in memory and making our candidate lists as small as possible in memory. For example zoekt has strategies of building up large lists, then score, sort, truncate. So that large list needs to be efficient.

I think in practice this is probably fine, with streaming we pay the cost of Content on the heap which should dominate?

Reply:
> Or is this for just ergonomics.

This one. We can get all the info we need from just start offset/end offset, but it's a bunch of fiddly rune counting and newline counting that was easier to do with the info zoekt has. That said, not paying for extra memory and network usage here is probably worth it

Reply:
I think the switch to uint32 negates my concerns here. We could do better here. I think in the case of large result sets your assumptions are low (in particular ranges per chunk). But from playing around with numbers we should be in a good place.